File: apps/backend/src/modules/users/users.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Put,
  Query,
  UseGuards,
} from '@nestjs/common';
import { JwtAuthGuard } from '@/common/guards/jwt.guard';
import { CurrentUser } from '@/common/decorators/user.decorator';
import { JwtPayload } from '@/common/types/jwt-payload';
import { UsersService } from './users.service';
import { UpdateUserSettingsDto } from './dto/user-settings.dto';
import { FollowStateDto } from './dto/follow-state.dto';
import { UserDetailsDto } from './dto/user-details.dto';
import { UpdateEmailDto } from './dto/update-email.dto';
import { UpdateHandleDto } from './dto/update-handle.dto';
import { AddMutedWordDto } from './dto/add-muted-word.dto';
import { RateLimitGuard } from '@/security/rate-limit.guard';

@Controller('users')
export class UsersController {
  constructor(private readonly users: UsersService) {}

  @UseGuards(JwtAuthGuard)
  @Get('me')
  me(@CurrentUser() user: JwtPayload) {
    return {
      userId: user.sub,
    };
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/account')
  account(@CurrentUser() user: JwtPayload) {
    return this.users.getAccountInfo(user.sub);
  }

  @UseGuards(JwtAuthGuard)
  @Put('me/email')
  updateEmail(
    @CurrentUser() user: JwtPayload,
    @Body() dto: UpdateEmailDto,
  ) {
    return this.users.updateEmail(user.sub, dto.email);
  }

  @UseGuards(JwtAuthGuard)
  @Put('me/handle')
  updateHandle(
    @CurrentUser() user: JwtPayload,
    @Body() dto: UpdateHandleDto,
  ) {
    return this.users.updateHandle(user.sub, dto);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/settings')
  getSettings(@CurrentUser() user: JwtPayload) {
    return this.users.getSettings(user.sub);
  }

  @UseGuards(JwtAuthGuard)
  @Put('me/settings')
  updateSettings(
    @CurrentUser() user: JwtPayload,
    @Body() dto: UpdateUserSettingsDto,
  ) {
    return this.users.updateSettings(user.sub, dto);
  }

  @UseGuards(JwtAuthGuard)
  @Put('me/details')
  updateDetails(
    @CurrentUser() user: JwtPayload,
    @Body() dto: UserDetailsDto,
  ) {
    return this.users.updateDetails(user.sub, dto);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/profile')
  profile(
    @CurrentUser() user: JwtPayload,
    @Query('limit') limit?: string,
  ) {
    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.getProfileSummary({
      userId: user.sub,
      limit: parsedLimit,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/blocks')
  blocked(
    @CurrentUser() user: JwtPayload,
    @Query('limit') limit?: string,
  ) {
    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.listBlockedUsers({
      userId: user.sub,
      limit: parsedLimit,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Post(':id/block')
  block(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
  ) {
    return this.users.blockUser({
      userId: user.sub,
      targetUserId,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete(':id/block')
  unblock(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
  ) {
    return this.users.unblockUser({
      userId: user.sub,
      targetUserId,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/muted-words')
  mutedWords(
    @CurrentUser() user: JwtPayload,
    @Query('limit') limit?: string,
  ) {
    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.listMutedWords({
      userId: user.sub,
      limit: parsedLimit,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Post('me/muted-words')
  addMutedWord(
    @CurrentUser() user: JwtPayload,
    @Body() dto: AddMutedWordDto,
  ) {
    return this.users.addMutedWord({
      userId: user.sub,
      phrase: dto.phrase,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete('me/muted-words/:id')
  removeMutedWord(
    @CurrentUser() user: JwtPayload,
    @Param('id') wordId: string,
  ) {
    return this.users.removeMutedWord({
      userId: user.sub,
      wordId,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Post('me/data-export')
  exportData(@CurrentUser() user: JwtPayload) {
    return this.users.createDataExport(user.sub);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/data-export')
  latestExport(@CurrentUser() user: JwtPayload) {
    return this.users.getLatestDataExport(user.sub);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me/connections')
  connections(
    @CurrentUser() user: JwtPayload,
    @Query('limit') limit?: string,
  ) {
    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.getConnections({
      userId: user.sub,
      limit: parsedLimit,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete('me')
  deleteAccount(@CurrentUser() user: JwtPayload) {
    return this.users.deleteAccount(user.sub);
  }

  @UseGuards(JwtAuthGuard)
  @Get(':id/profile')
  publicProfile(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
    @Query('limit') limit?: string,
  ) {
    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.getPublicProfileSummary({
      targetUserId,
      viewerId: user.sub,
      limit: parsedLimit,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Get('search')
  search(
    @CurrentUser() user: JwtPayload,
    @Query('query') query?: string,
    @Query('limit') limit?: string,
  ) {
    if (!query) {
      throw new BadRequestException('query is required');
    }

    const parsedLimit =
      limit && Number.isFinite(Number(limit))
        ? Number(limit)
        : undefined;

    return this.users.searchUsers({
      userId: user.sub,
      query,
      limit: parsedLimit,
    });
  }

  @UseGuards(RateLimitGuard)
  @Get('username-available')
  async usernameAvailable(@Query('username') username?: string) {
    if (!username) {
      throw new BadRequestException('username is required');
    }

    const available = await this.users.isUsernameAvailable(username);
    return { available };
  }

  @UseGuards(JwtAuthGuard)
  @Post(':id/follow')
  toggleFollow(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
  ) {
    return this.users.toggleFollow({
      followerId: user.sub,
      followingId: targetUserId,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Put(':id/follow')
  setFollow(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
    @Body() dto: FollowStateDto,
  ) {
    return this.users.setFollow({
      followerId: user.sub,
      followingId: targetUserId,
      follow: dto.follow,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete(':id/follower')
  removeFollower(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
  ) {
    return this.users.removeFollower({
      userId: user.sub,
      followerId: targetUserId,
    });
  }

  @UseGuards(JwtAuthGuard)
  @Delete(':id/connection')
  removeConnection(
    @CurrentUser() user: JwtPayload,
    @Param('id') targetUserId: string,
  ) {
    return this.users.removeConnection({
      userId: user.sub,
      targetUserId,
    });
  }
}

File: apps/backend/src/modules/users/users_controller.cpp
#include "modules/users/users_controller.h"

#include <cstdlib>
#include <optional>
#include <string>
#include <unordered_set>

#include "app_state.h"
#include "http/json.h"
#include "http/response.h"
#include "modules/users/users_service.h"

namespace {

bool ParseJsonPayload(const drogon::HttpRequestPtr& req,
                      const std::unordered_set<std::string>& allowed,
                      Json::Value& out,
                      drogon::HttpResponsePtr& error_resp) {
  std::string error;
  if (!http::ParseJsonObject(req, out, error)) {
    error_resp = http::ErrorResponse(drogon::k400BadRequest, error);
    return false;
  }
  if (!http::HasOnlyFields(out, allowed)) {
    error_resp = http::ErrorResponse(drogon::k400BadRequest, "Invalid payload");
    return false;
  }
  return true;
}

bool GetRequiredString(const Json::Value& body,
                       const std::string& key,
                       std::string& out) {
  return http::GetStringField(body, key, out);
}

bool GetOptionalString(const Json::Value& body,
                       const std::string& key,
                       std::optional<std::string>& out) {
  if (!body.isMember(key)) {
    return true;
  }
  std::string value;
  if (!http::GetStringField(body, key, value)) {
    return false;
  }
  out = value;
  return true;
}

bool GetRequiredBool(const Json::Value& body,
                     const std::string& key,
                     bool& out) {
  if (!body.isMember(key)) {
    return false;
  }
  const auto& value = body[key];
  if (!value.isBool()) {
    return false;
  }
  out = value.asBool();
  return true;
}

bool SetOptionalBool(const Json::Value& body,
                     const std::string& key,
                     Json::Value& updates) {
  if (!body.isMember(key)) {
    return true;
  }
  const auto& value = body[key];
  if (!value.isBool()) {
    return false;
  }
  updates[key] = value.asBool();
  return true;
}

bool SetOptionalInt(const Json::Value& body,
                    const std::string& key,
                    Json::Value& updates) {
  if (!body.isMember(key)) {
    return true;
  }
  const auto& value = body[key];
  if (!value.isInt()) {
    return false;
  }
  updates[key] = value.asInt();
  return true;
}

bool SetOptionalNumber(const Json::Value& body,
                       const std::string& key,
                       Json::Value& updates) {
  if (!body.isMember(key)) {
    return true;
  }
  const auto& value = body[key];
  if (!value.isNumeric()) {
    return false;
  }
  updates[key] = value.asDouble();
  return true;
}

bool SetOptionalString(const Json::Value& body,
                       const std::string& key,
                       Json::Value& updates) {
  if (!body.isMember(key)) {
    return true;
  }
  const auto& value = body[key];
  if (!value.isString()) {
    return false;
  }
  updates[key] = value.asString();
  return true;
}

bool GetUserId(const drogon::HttpRequestPtr& req, std::string& user_id) {
  const auto attrs = req->getAttributes();
  if (!attrs || !attrs->find("user_id")) {
    return false;
  }
  user_id = attrs->get<std::string>("user_id");
  return true;
}

std::optional<int> ParseOptionalInt(const drogon::HttpRequestPtr& req,
                                    const std::string& key) {
  const std::string value = req->getParameter(key);
  if (value.empty()) {
    return std::nullopt;
  }

  char* end = nullptr;
  const long parsed = std::strtol(value.c_str(), &end, 10);
  if (end == value.c_str() || *end != '\0') {
    return std::nullopt;
  }

  return static_cast<int>(parsed);
}

UsersService BuildUsersService() {
  return UsersService(AppState::Instance().GetDb());
}

void RespondWithUsers(
    std::function<Json::Value()> handler,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) {
  try {
    const Json::Value payload = handler();
    callback(http::JsonResponse(payload, drogon::k200OK));
  } catch (const UsersError& err) {
    callback(http::ErrorResponse(err.status, err.what()));
  } catch (const std::exception&) {
    callback(http::ErrorResponse(drogon::k500InternalServerError,
                                 "Internal server error"));
  }
}

}  // namespace

void UsersController::Me(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  Json::Value payload;
  payload["userId"] = user_id;
  callback(http::JsonResponse(payload, drogon::k200OK));
}

void UsersController::Account(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id]() {
        auto users = BuildUsersService();
        return users.GetAccountInfo(user_id);
      },
      std::move(callback));
}

void UsersController::UpdateEmail(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  static const std::unordered_set<std::string> allowed = {"email"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  std::string email;
  if (!GetRequiredString(body, "email", email) || email.size() > 255) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id, &email]() {
        auto users = BuildUsersService();
        return users.UpdateEmail(user_id, email);
      },
      std::move(callback));
}

void UsersController::UpdateHandle(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  static const std::unordered_set<std::string> allowed = {"username",
                                                          "displayName",
                                                          "bio",
                                                          "location",
                                                          "website"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  UpdateHandleInput input;
  if (!GetOptionalString(body, "username", input.username) ||
      !GetOptionalString(body, "displayName", input.display_name) ||
      !GetOptionalString(body, "bio", input.bio) ||
      !GetOptionalString(body, "location", input.location) ||
      !GetOptionalString(body, "website", input.website)) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  if (input.username &&
      (input.username->size() < 3 || input.username->size() > 32)) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }
  if (input.display_name && input.display_name->size() > 64) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }
  if (input.bio && input.bio->size() > 160) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }
  if (input.location && input.location->size() > 120) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }
  if (input.website && input.website->size() > 255) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id, &input]() {
        auto users = BuildUsersService();
        return users.UpdateHandle(user_id, input);
      },
      std::move(callback));
}

void UsersController::GetSettings(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id]() {
        auto users = BuildUsersService();
        return users.GetSettings(user_id);
      },
      std::move(callback));
}

void UsersController::UpdateSettings(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  static const std::unordered_set<std::string> allowed = {
      "privateAccount",
      "activityStatus",
      "readReceipts",
      "messagePreview",
      "sensitiveContent",
      "locationSharing",
      "twoFactor",
      "loginAlerts",
      "appLock",
      "biometrics",
      "pushNotifications",
      "emailNotifications",
      "inAppSounds",
      "inAppHaptics",
      "dataSaver",
      "autoDownload",
      "autoPlayVideos",
      "reduceMotion",
      "themeIndex",
      "textScale",
      "languageLabel"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  Json::Value updates(Json::objectValue);
  if (!SetOptionalBool(body, "privateAccount", updates) ||
      !SetOptionalBool(body, "activityStatus", updates) ||
      !SetOptionalBool(body, "readReceipts", updates) ||
      !SetOptionalBool(body, "messagePreview", updates) ||
      !SetOptionalBool(body, "sensitiveContent", updates) ||
      !SetOptionalBool(body, "locationSharing", updates) ||
      !SetOptionalBool(body, "twoFactor", updates) ||
      !SetOptionalBool(body, "loginAlerts", updates) ||
      !SetOptionalBool(body, "appLock", updates) ||
      !SetOptionalBool(body, "biometrics", updates) ||
      !SetOptionalBool(body, "pushNotifications", updates) ||
      !SetOptionalBool(body, "emailNotifications", updates) ||
      !SetOptionalBool(body, "inAppSounds", updates) ||
      !SetOptionalBool(body, "inAppHaptics", updates) ||
      !SetOptionalBool(body, "dataSaver", updates) ||
      !SetOptionalBool(body, "autoDownload", updates) ||
      !SetOptionalBool(body, "autoPlayVideos", updates) ||
      !SetOptionalBool(body, "reduceMotion", updates) ||
      !SetOptionalInt(body, "themeIndex", updates) ||
      !SetOptionalNumber(body, "textScale", updates) ||
      !SetOptionalString(body, "languageLabel", updates)) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id, &updates]() {
        auto users = BuildUsersService();
        return users.UpdateSettings(user_id, updates);
      },
      std::move(callback));
}

void UsersController::UpdateDetails(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  static const std::unordered_set<std::string> allowed = {"firstName",
                                                          "lastName",
                                                          "phoneCountryCode",
                                                          "phoneNumber"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  UpdateDetailsInput input;
  if (!GetRequiredString(body, "firstName", input.first_name) ||
      !GetRequiredString(body, "lastName", input.last_name) ||
      !GetRequiredString(body, "phoneCountryCode",
                         input.phone_country_code) ||
      !GetRequiredString(body, "phoneNumber", input.phone_number)) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id, &input]() {
        auto users = BuildUsersService();
        return users.UpdateDetails(user_id, input);
      },
      std::move(callback));
}

void UsersController::Profile(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  UserLimitInput input{user_id, ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.GetProfileSummary(input);
      },
      std::move(callback));
}

void UsersController::Blocked(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  UserLimitInput input{user_id, ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.ListBlockedUsers(input);
      },
      std::move(callback));
}

void UsersController::BlockUser(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  BlockInput input{user_id, target_user_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.BlockUser(input);
      },
      std::move(callback));
}

void UsersController::UnblockUser(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  BlockInput input{user_id, target_user_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.UnblockUser(input);
      },
      std::move(callback));
}

void UsersController::MutedWords(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  UserLimitInput input{user_id, ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.ListMutedWords(input);
      },
      std::move(callback));
}

void UsersController::AddMutedWord(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  static const std::unordered_set<std::string> allowed = {"phrase"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  std::string phrase;
  if (!GetRequiredString(body, "phrase", phrase) || phrase.empty() ||
      phrase.size() > 120) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  AddMutedWordInput input{user_id, phrase};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.AddMutedWord(input);
      },
      std::move(callback));
}

void UsersController::RemoveMutedWord(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& word_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RemoveMutedWordInput input{user_id, word_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.RemoveMutedWord(input);
      },
      std::move(callback));
}

void UsersController::ExportData(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id]() {
        auto users = BuildUsersService();
        return users.CreateDataExport(user_id);
      },
      std::move(callback));
}

void UsersController::LatestExport(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id]() {
        auto users = BuildUsersService();
        return users.GetLatestDataExport(user_id);
      },
      std::move(callback));
}

void UsersController::Connections(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  UserLimitInput input{user_id, ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.GetConnections(input);
      },
      std::move(callback));
}

void UsersController::DeleteAccount(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RespondWithUsers(
      [&user_id]() {
        auto users = BuildUsersService();
        return users.DeleteAccount(user_id);
      },
      std::move(callback));
}

void UsersController::PublicProfile(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  PublicProfileInput input{target_user_id, user_id,
                           ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.GetPublicProfileSummary(input);
      },
      std::move(callback));
}

void UsersController::Search(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  const std::string query = req->getParameter("query");
  if (query.empty()) {
    callback(
        http::ErrorResponse(drogon::k400BadRequest, "query is required"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  SearchUsersInput input{user_id, query, ParseOptionalInt(req, "limit")};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.SearchUsers(input);
      },
      std::move(callback));
}

void UsersController::UsernameAvailable(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback) const {
  const std::string username = req->getParameter("username");
  if (username.empty()) {
    callback(http::ErrorResponse(drogon::k400BadRequest,
                                 "username is required"));
    return;
  }

  RespondWithUsers(
      [&username]() {
        auto users = BuildUsersService();
        Json::Value payload;
        payload["available"] = users.IsUsernameAvailable(username);
        return payload;
      },
      std::move(callback));
}

void UsersController::ToggleFollow(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  FollowInput input{user_id, target_user_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.ToggleFollow(input);
      },
      std::move(callback));
}

void UsersController::SetFollow(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  static const std::unordered_set<std::string> allowed = {"follow"};

  Json::Value body;
  drogon::HttpResponsePtr error_resp;
  if (!ParseJsonPayload(req, allowed, body, error_resp)) {
    callback(error_resp);
    return;
  }

  bool follow = false;
  if (!GetRequiredBool(body, "follow", follow)) {
    callback(http::ErrorResponse(drogon::k400BadRequest, "Invalid payload"));
    return;
  }

  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  SetFollowInput input{user_id, target_user_id, follow};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.SetFollow(input);
      },
      std::move(callback));
}

void UsersController::RemoveFollower(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RemoveFollowerInput input{user_id, target_user_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.RemoveFollower(input);
      },
      std::move(callback));
}

void UsersController::RemoveConnection(
    const drogon::HttpRequestPtr& req,
    std::function<void(const drogon::HttpResponsePtr&)>&& callback,
    const std::string& target_user_id) const {
  std::string user_id;
  if (!GetUserId(req, user_id)) {
    callback(http::ErrorResponse(drogon::k401Unauthorized, "unauthorized"));
    return;
  }

  RemoveConnectionInput input{user_id, target_user_id};

  RespondWithUsers(
      [&input]() {
        auto users = BuildUsersService();
        return users.RemoveConnection(input);
      },
      std::move(callback));
}
